File: ts/packages/libs/src/libs/jwt/jwt.ts
import BaseLib from '../../base.js';
import jsonWebToken from 'jsonwebtoken';
import { type SignOptions } from 'jsonwebtoken';

class JwtLib extends BaseLib {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	async create(payload: any, expiresIn: SignOptions['expiresIn'], JWT_SECRET: string) {
		return jsonWebToken.sign(payload, JWT_SECRET, { expiresIn: expiresIn });
	}

	async verify(token: string, JWT_SECRET: string) {
		return jsonWebToken.verify(token, JWT_SECRET);
	}

	getExpires() {
		const time = this.getExpiresTime();
		const atTime = this.getExpiresAtTime(time);
		const expires: JWTExpires = {
			time,
			atTime,
		};
		return expires;
	}

	private getExpiresAtTime(expiresTime: number) {
		return new Date(Date.now() + expiresTime);
	}

	private getExpiresTime() {
		return 15 * 60 * 1000;
	}
}

export type JWTExpires = {
	time: number;
	atTime: Date;
};

export default JwtLib;

File: dotnet/Packages/Libs/Tokens/Jwt.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.IdentityModel.Tokens;

namespace zed31rus.Packages.Libs.Tokens;

public record JwtExpires(TimeSpan Time, DateTime AtTime);

public interface IJwt
{
    JwtExpires GetExpires();
    string Create(IEnumerable<Claim> claims, TimeSpan expiresIn, string jwtSecret);
    ClaimsPrincipal Verify(string token, string jwtSecret);
}

internal class Jwt : IJwt
{
    private readonly JwtSecurityTokenHandler _tokenHandler = new();

    private TimeSpan GetExpiresTime()
    {
        return TimeSpan.FromMinutes(15);
    }

    private DateTime GetExpiresAtTime(TimeSpan expiresTime)
    {
        return DateTime.UtcNow.Add(expiresTime);
    }

    public JwtExpires GetExpires()
    {
        var time = GetExpiresTime();
        var atTime = GetExpiresAtTime(time);
        return new JwtExpires(time, atTime);
    }

    public string Create(IEnumerable<Claim> claims, TimeSpan expiresIn, string jwtSecret)
    {
        var key = Encoding.UTF8.GetBytes(jwtSecret);

        var tokenDescriptor = new SecurityTokenDescriptor
        {
            Subject = new ClaimsIdentity(claims),
            Expires = DateTime.UtcNow.Add(expiresIn),
            SigningCredentials = new SigningCredentials(
                new SymmetricSecurityKey(key),
                SecurityAlgorithms.HmacSha256Signature
            )
        };

        var token = _tokenHandler.CreateToken(tokenDescriptor);
        return _tokenHandler.WriteToken(token);
    }

    public ClaimsPrincipal Verify(string token, string jwtSecret)
    {
        var key = Encoding.UTF8.GetBytes(jwtSecret);

        var validationParameters = new TokenValidationParameters
        {
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = new SymmetricSecurityKey(key),
            ValidateIssuer = false,
            ValidateAudience = false,
            ValidateLifetime = true,
            ClockSkew = TimeSpan.Zero
        };

        var principal = _tokenHandler.ValidateToken(token, validationParameters, out _);
        return principal;
    }
}